pkg/extension/create_interceptor.go

```diff
@@ -0,0 +1,160 @@
+
+package extension
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/Peripli/service-manager/storage"
+)
+
+type namedCreateAPIFunc struct {
+	Name string
+	Func func(InterceptCreateOnAPI) InterceptCreateOnAPI
+}
+
+type namedCreateTxFunc struct {
+	Name string
+	Func func(InterceptCreateOnTx) InterceptCreateOnTx
+}
+
+type createHookOnAPIHandler struct {
+	CreateHookOnAPIFuncs         []*namedCreateAPIFunc
+	CreateHookOnTransactionFuncs []*namedCreateTxFunc
+}
+
+func (c *createHookOnAPIHandler) OnAPICreate(f InterceptCreateOnAPI) InterceptCreateOnAPI {
+	for i := range c.CreateHookOnAPIFuncs {
+		f = c.CreateHookOnAPIFuncs[len(c.CreateHookOnAPIFuncs)-1-i].Func(f)
+	}
+	return f
+}
+
+func (c *createHookOnAPIHandler) OnTxCreate(f InterceptCreateOnTx) InterceptCreateOnTx {
+	for i := range c.CreateHookOnTransactionFuncs {
+		f = c.CreateHookOnTransactionFuncs[len(c.CreateHookOnTransactionFuncs)-1-i].Func(f)
+	}
+	return f
+}
+
+// UnionCreateInterceptor returns a function which spawns all create interceptors, sorts them and wraps them into one.
+func UnionCreateInterceptor(providers []CreateInterceptorProvider) func() CreateInterceptor {
+	return func() CreateInterceptor {
+		c := &createHookOnAPIHandler{}
+		c.CreateHookOnAPIFuncs = make([]*namedCreateAPIFunc, 0, len(providers))
+		c.CreateHookOnTransactionFuncs = make([]*namedCreateTxFunc, 0, len(providers))
+
+		for _, p := range providers {
+			hook := p.Provide()
+			positionAPIType := PositionNone
+			positionTxType := PositionNone
+			nameAPI := ""
+			nameTx := ""
+
+			if orderedProvider, isOrdered := p.(Ordered); isOrdered {
+				positionAPIType, nameAPI = orderedProvider.PositionAPI()
+				positionTxType, nameTx = orderedProvider.PositionTx()
+			}
+
+			c.insertAPIFunc(positionAPIType, nameAPI, &namedCreateAPIFunc{
+				Name: p.Name(),
+				Func: hook.OnAPICreate,
+			})
+			c.insertTxFunc(positionTxType, nameTx, &namedCreateTxFunc{
+				Name: p.Name(),
+				Func: hook.OnTxCreate,
+			})
+		}
+		return c
+	}
+}
+
+// CreateInterceptorProvider provides CreateInterceptors for each request
+//go:generate counterfeiter . CreateInterceptorProvider
+type CreateInterceptorProvider interface {
+	Named
+	Provide() CreateInterceptor
+}
+
+// InterceptCreateOnAPI hook for entity creation outside of transaction
+type InterceptCreateOnAPI func(ctx context.Context, obj types.Object) (types.Object, error)
+
+// InterceptCreateOnTx hook for entity creation in transaction
+type InterceptCreateOnTx func(ctx context.Context, txStorage storage.Warehouse, newObject types.Object) error
+
+// CreateInterceptor provides hooks on entity creation
+//go:generate counterfeiter . CreateInterceptor
+type CreateInterceptor interface {
+	OnAPICreate(h InterceptCreateOnAPI) InterceptCreateOnAPI
+	OnTxCreate(f InterceptCreateOnTx) InterceptCreateOnTx
+}
+
+func (c *createHookOnAPIHandler) insertAPIFunc(positionType PositionType, name string, h *namedCreateAPIFunc) {
```

Comment:
just pass separately the funcName and the func (removing the wrapper utility type